### DISCUSSION TITLE
How can I know if a plugin is enabled in another plugin?

### DISCUSSION THREAD

**Discussion by Mister-Hope** [vuepress/core#193] [Q&A] [answered]
My basic question is, how can a plugin know if the plugin it want to use is enabled or not.

e.g.: my sitemap and seo plugin would both want git plugin to be enabled.

If I redeclare plugins to make sure they are enabled, I will get errors like 

```
warning plugin @vuepress/plugin-git has been used multiple times, only the last one will take effect
```

And this can be awful when I am using:

```
git → reading-time→   theme
↓            ↓       ↗ ↑
  ↘   →  pageinfo      ↑
      ↘   →   seo  ↗  ↑
           ↘ sitemap  ↗
              ↘ feed↗



     ↗        →      ↘
palatte→ reading-time→   theme
↓            ↓       ↗ ↑
  ↘   →  pageinfo      ↑
      ↘   →   seo  ↗  ↑
           ↘ sitemap  ↗
              ↘ feed↗
           ↘copy-code↗
    ↘md-enhance↗
       ↘  pwa↗

(palette is my own plugin controling styles)
```

when I want them to be able to work standalone and in one theme

Common idea are:

-  check in my plugin whether the plugin has already been declared or not and decide whether to enable it. 

   This seems hard to implement

-  Make a soft declare on depending on this plugin and wait for users to input options.

     In this way only theme and user config have full control of using and setting options for plugins, while plugins can only decalre the plugins they are depending and emit warnings if the plugin is not actually enabled. This is a huge breaking change

-  Make a search squence, and then merge all the options if the plugin(do not support mutiple instance) is declared in mutiple plugins or theme and warned if some options are covered when merging.
 
   This don't have much affect to current code.

I hope one of them above can be achieved, or there is something which I can do to avoid the warning.

**Comment by meteorlxy** [comment, marked as answer]
You can use following snippet to detect if a plugin has already been used:

https://github.com/vuepress/vuepress-next/blob/5e96641b8d2dd947372222e0af2ea47729afbada/packages/%40vuepress/core/src/app/appUse.ts#L17-L32

---

TBH, supporting nested plugins does have this kind of tradeoff.

> @Shigma made [vuepress-mergable](https://vuepress-community.netlify.app/en/tools/mergeable/) in v1 to solve this problem (I think it wouldn't work in v2 now 😓 ).

**Reply by Mister-Hope** [reply]
I think for quite a while, and think maybe we can support this:

```
['plugin', options, (options)=>{
  // change current options
}]
```

- If the plugin has been enabled previously, use the third function and pass in the options(if not supporting muti), and use the return value as options.

- If not, just use the second option and behave normaly

Also in the new api, `app.pluginApi.plugins` will held both the resolved plugin object and options, and the item should be updated if a later plugin config is having the third function option.

So e.g. in git plugin:

A plugin called A will only need to read contributors. so he can use:

```js
['A', 
  { contributors: true, createdTime: false, updatedTime: false },
 ( options) => ({ ...(options || {}), contributors: true,)
]
```

Which will not break the existing behaviour while not enabling features uneeded.

---

### Why not  something like `vuepress-mergable`?

The previous option is already lost in the `app.pluginApi.plugins`. Developer have no idea what do the previous option look like.

E.g:

```js
(options) => {
  name: 'x',
  onGenerated() {
     // reads option
  }
}
```

The options cannot be inffered from the resolved pluginObject in `app.pluginApi.plugins`

---

### Other issues

From your code, I am afriaid that if users are using:

```
[
  ['a', options],
  ['b', options]
]
```

while `b`'s plugin chain is using `a`, users options will be lost.


---

@meteorlxy Any idea about this?


If you think the third option is too complicated for configuation, I still would like to reshape app.pluginApi` to held both resolved plugin api and the original options. That would at least make developers read the existing plugin options and update it.

**Reply by Mister-Hope** [reply]
## Additional info: Why am I so concerned about it?

VuePress 2 is fully decoupled with plugins. And that means some 'tool-like' plugins may used everywhere.

Like common options  `hostname` `author` `logo` may be used in server plugins, A lot of commnuity plugins are having a fallback to `themeConfig.xxx` besides their options because users really don't want to decalre same value in each plugin options.So these plugins will all use `plugin-theme-data`.

git-plugin can be another example where feed seo sitemap pageinfo plugins will depend on it.

Also something like your palette plugin which allow plugins to read config to change styles will also be commonly used.

Because there are things like breakpoints, dark selecter and so on cannot be configured through css vars.

In this situation, a good handing of dumplicated plugin usage is important in my opinion.

**Reply by meteorlxy** [reply]
That's really a good concern.

Supporting nested plugins is convenient for plugin authors, but it is also kind of a bug hole 😓 .

---

`vuepress-mergable` (called **Solution A**) and the solution you proposed (called **Solution B**) are trying to solve this problem in different way.

- **Solution A**: Store the previous options in a plugin closure, and handle the merge strategy by plugin itself.

- **Solution B**: Store the previous options in core / pluginApi, and handle the merge strategy by plugin users. (As I remember, Shigma had ever proposed a solution like this, which was declined by Ulivz, so he created `vuepress-mergable` to solve the problem in plugin side)

**Reply by Mister-Hope** [reply]
Also the priority should be dicussed and enahcned, I do think that:

Users' plugin config > theme config > plugin tool chain from above and below.

### So:

User:

- A1
- B1
  - A2
  - D1
- C

Theme:

- A2
- D2
- E
   - D3
     - A3

Should be read as:

- A3
- D3
- A2
- D2
- E
- A2
- D1
- A1
- B1
- C

The deeper, the weaker.

And unlike V1, I think it should be supported by core.

**Reply by meteorlxy** [reply]
I had ever considered to drop support for nested plugins, i.e. only user-config and theme can use plugins, while plugins cannot use other plugins.

It could be an important breaking change for plugin authors, so I didn't drop it at that time.

That could be one of the solution of this problem, and I think it might not break a lot.

If a plugin is depending on other plugins, we can put them in the `peerDependencies`, and tell users to use them together in the docs (just like `@vuepress/pwa` and `@vuepress/pwa-popup`).

WDYT? 🤔

**Reply by Mister-Hope** [reply]
I think making a soft declare should be better.

Plugins should declare what plugins they are depending, and if the plugins are not enabled finally, a warning should be thrown by @vuepress/core.

That means @vuepress/core should check if the plugins in declarations are all in use when the plugin and theme are all loaded.

I do think this is better, For my users, 1/3 of them won't carefully read the docs.

**Reply by Mister-Hope** [reply]
Also I think we have to warn the developers, they should only declare plugins which theirs are hardly depend on, if a plugin is optional, then it shouldn't be declared.

**Reply by Mister-Hope** [reply]
I have another reason. why soft declaration should exist:

We are not displaying what plugins we use now, so users won't know what plugins they are actully using expect those they declared themselves.

Users are not concerned about what the tool chain like, especially for plugins in themes.

**Reply by Mister-Hope** [reply]
Wait, that's not good for plugins that have mutiple instance (something like `@vuepress/plugin-container` commonly used in markdown plugins and something like diy palette plugin)

So maybe:

- Plugins can only use other plugins which support `mutiple`
- Plugins should declare plugins they are depending at, and a warn should be displayed if the plugins is not actually in use

The above two are different options. Maybe (`plugins` and `requires`)

WDYT?😉

**Reply by meteorlxy** [reply]
A plugin can print warning when some deps are missing.

I prefer not to add options if it is not necessary (the reason why `clientDynamicModules` was removed).

**Reply by Mister-Hope** [reply]
Actually no new options are need to add, just using the existing plugin option and make:

- Actual use plugin supporring mutiple
- Check and warn if the other plugin listed not in use.

**Reply by Mister-Hope** [reply]
I saw you remove the entire plugin config in the latest commit.

But I think supporting plugins to use `mutiple` plugin is good, I have common cases to use some 'tool-like' plugins in my other plugins e.g.: a palette system.

It's hard to make a `mutiple` plugin used in other plugins just like vue3's composition api.@meteorlxy Letting users configue them mutiple times will lead to a confused plugin config.

> Check and warn if the other plugin listed not in use.

If you think above is unecessary, I still want you to rethink about this: 

> - can use plugin which supports `mutiple`

This has no side effects



<div class="TimelineItem js-comment-container js-targetable-element ml-0 py-2 pl-3 discussion-nested-comment-timeline-item discussion-primer-next-nested-comment-timeline-item" id="discussioncomment-843357" style="box-sizing: border-box; margin-left: 16px; padding: 16px 0px; position: relative; display: flex; color: rgb(201, 209, 217); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="discussions-timeline-scroll-target" id="discussioncomment-843357" data-gid="MDE3OkRpc2N1c3Npb25Db21tZW50ODQzMzU3" data-url="/vuepress/vuepress-next/discussions/193/comments/843357" data-error="" style="box-sizing: border-box; width: 828px; padding-top: 60px; margin-top: -60px; pointer-events: none !important;"><div class="
      timeline-comment-group
      discussion-comment
      js-discussion-comment
      js-minimizable-comment-group
      js-targetable-element
      TimelineItem-body
      discussion-nested-comment-group position-relative
      mt-0 pl-2
    " style="box-sizing: border-box; position: relative !important; margin-top: 4px; padding-left: 8px !important; min-width: 0px; max-width: 100%; color: var(--color-timeline-text); flex: 1 1 auto; pointer-events: auto; margin-left: 32px;"><div class="d-flex" style="box-sizing: border-box; display: flex !important;"><div class="width-full" style="box-sizing: border-box; width: 788px;"><div class="timeline-comment unminimized-comment comment previewable-edit js-task-list-container flex-auto box-shadow-none js-comment current-user reorderable-task-lists border-0 color-bg-canvas-inset nested-discussion-timeline-comment ml-n3" data-body-version="ef5a60451e8a2ff4d30e1c36b7c008ce52748318a395f4fd006a01cd9b6dbf90" data-unfurl-hide-url="/content_reference_attachments/hide" style="box-sizing: border-box; border-top-width: ; border-right-width: ; border-bottom-width: ; border-left-width: ; border-top-style: ; border-right-style: ; border-bottom-style: ; border-left-style: ; border-color: var(--color-box-border-info); border-image-source: ; border-image-slice: ; border-image-width: ; border-image-outset: ; border-image-repeat: ; box-shadow: none !important; background-color: var(--color-bg-primary); flex: 1 1 auto !important; margin-left: -16px !important; position: relative; color: var(--color-text-primary); border-radius: 6px;"><div class="" style="box-sizing: border-box;"><div class="edit-comment-hide" style="box-sizing: border-box;"><task-lists sortable="" style="box-sizing: border-box;">Actual use plugin supporring mutipleCheck and warn if the other plugin listed not in use.</task-lists><div class="d-flex flex-justify-between flex-content-center" style="box-sizing: border-box; place-content: center space-between !important; display: flex !important;"><div class="comment-reactions d-flex flex-items-start flex-row-reverse js-reactions-container js-reaction-buttons-container flex-justify-end" style="box-sizing: border-box; flex-direction: row-reverse !important; justify-content: flex-end !important; align-items: flex-start !important; display: none; margin-left: 16px; border-top: 0px !important;"><form class="js-pick-reaction" action="https://github.com/vuepress/vuepress-next/reactions" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"><div class="comment-reactions-options js-comment-reactions-options" style="box-sizing: border-box;"></div></form></div></div></div><form class="js-comment-update" action="https://github.com/vuepress/vuepress-next/discussions/193/comments/843357" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"><div class="js-edit-comment-form-container" data-body-version="ef5a60451e8a2ff4d30e1c36b7c008ce52748318a395f4fd006a01cd9b6dbf90" data-outer-div-classes="border-md mx-3 color-bg-primary rounded-1" data-permissions-url="/vuepress/vuepress-next/discussions/193/comments/843357/permissions" data-repository-id="261759626" data-target-body="Actually no new options are need to add, just using the existing plugin option and make:

- Actual use plugin supporring mutiple
- Check and warn if the other plugin listed not in use." data-target-type="comment" data-textarea-id="discussioncomment-843357-body" data-slash-command-url="/vuepress/vuepress-next/slash_apps?subject_gid=MDEwOkRpc2N1c3Npb24zNDAxNDkx" style="box-sizing: border-box;"></div></form></div></div></div></div></div></div></div><div class="TimelineItem js-comment-container js-targetable-element ml-0 py-2 pl-3 discussion-nested-comment-timeline-item discussion-primer-next-nested-comment-timeline-item" id="discussioncomment-844867" style="box-sizing: border-box; margin-left: 16px; padding: 16px 0px; position: relative; display: flex; color: rgb(201, 209, 217); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><br class="Apple-interchange-newline">

</div>

**Reply by meteorlxy** [reply]
In fact, you can still use plugins via `app.use()` in your plugin.

But it will be an undocumented usage, because we should discourage plugin nesting.

**Reply by Mister-Hope** [reply]
Got it, thanks!😀
